# Segmenting and Clustering Neighborhoods in Toronto
## Part Three

We will be using the BeautifulSoup package to scrape the table from the provided web address.

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [2]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

In [3]:
soup = BeautifulSoup(website_url, 'lxml')

In [4]:
table = soup.find_all('table', class_ = "wikitable sortable")

In [5]:
len(table)

1

Fortunately there is only one table on the page, meaning we do not have to narrow anything down to scrape our data.

In [6]:
table = table[0]
type(table)

bs4.element.Tag

Now it's time to convert the table to a dataframe.

In [7]:
table_rows = table.find_all('tr')

res = []
for tr in table_rows:
    td = tr.find_all('td')
    row = [tr.text.strip() for tr in td if tr.text.strip()]
    if row:
        res.append(row)
df = pd.DataFrame(res, columns=["Postcode", "Borough", "Neighbourhood"])
df=df.rename(columns = {'Postcode':'Postal Code', 'Neighbourhood':'Neighborhood'})
df.head(10)

Postal Code           Borough      Neighborhood
0         M1A      Not assigned      Not assigned
1         M2A      Not assigned      Not assigned
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M5A  Downtown Toronto       Regent Park
6         M6A        North York  Lawrence Heights
7         M6A        North York    Lawrence Manor
8         M7A      Queen's Park      Not assigned
9         M8A      Not assigned      Not assigned

In [8]:
df.shape

(288, 3)

First we'll remove the postal codes that don't have a borough.

In [9]:
df1 = df[df.Borough != 'Not assigned']
df1.head(10)

Postal Code           Borough      Neighborhood
2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park      Not assigned
10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern

In [10]:
df1.shape

(211, 3)

Then we'll assign the areas that don't have a neighborhood to their overarching borough.

In [11]:
df1.Neighborhood[df1.Neighborhood == 'Not assigned'] = df1.Borough
df1.head(10)

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/generic.py:8682: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)
/opt/conda/envs/Python36/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3267: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


Postal Code           Borough      Neighborhood
2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park      Queen's Park
10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern

Finally we'll group the neighborhoods by their postal code and borough.

In [12]:
df2 = df1.groupby(['Postal Code','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df2.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [13]:
df2.shape

(103, 3)

The dataframe has 103 rows and 3 columns. Now that the dataframe has been established and cleaned, we need to get the geospatial data for each postal code. We will be using the provided .csv file.

In [14]:
url="http://cocl.us/Geospatial_data"
df3=pd.read_csv(url)
df3.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Since the .csv file is easily read as a dataframe, all we have to do is merge it with our previous dataframe.

In [15]:
result = pd.merge(df2,df3, on='Postal Code')
result.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

With the dataframe now containing all the information we need, we can start analysis. First we'll install and import all the packages we'll need.

In [16]:
import numpy as np 

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

#!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize 


import matplotlib.cm as cm
import matplotlib.colors as colors


from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
print('Libraries imported.')

Libraries imported.


In [17]:
address = '1090 Don Mills Rd, North York, ON M3C 3R6, Canada'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of 1090 Don Mills Rd, North York, ON M3C 3R6, Canadao is {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.20.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of 1090 Don Mills Rd, North York, ON M3C 3R6, Canadao is 43.7362962, -79.343917.


In [18]:
map = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(result['Latitude'], result['Longitude'], result['Borough'], result['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map) 
map

Let's start by narrowing our analysis, focusing on North York.

In [19]:
NorthYork_data = result[result.Borough == 'North York']
NorthYork_data.head()

Postal Code     Borough                  Neighborhood   Latitude  Longitude
17         M2H  North York             Hillcrest Village  43.803762 -79.363452
18         M2J  North York  Fairview, Henry Farm, Oriole  43.778517 -79.346556
19         M2K  North York               Bayview Village  43.786947 -79.385975
20         M2L  North York      Silver Hills, York Mills  43.757490 -79.374714
21         M2M  North York       Newtonbrook, Willowdale  43.789053 -79.408493

In [20]:
NorthYork=NorthYork_data.reset_index()
NorthYork = NorthYork[NorthYork.Neighborhood != 'Silver Hills, York Mills']
NorthYork

index Postal Code     Borough  \
0      17         M2H  North York   
1      18         M2J  North York   
2      19         M2K  North York   
4      21         M2M  North York   
5      22         M2N  North York   
6      23         M2P  North York   
7      24         M2R  North York   
8      25         M3A  North York   
9      26         M3B  North York   
10     27         M3C  North York   
11     28         M3H  North York   
12     29         M3J  North York   
13     30         M3K  North York   
14     31         M3L  North York   
15     32         M3M  North York   
16     33         M3N  North York   
17     34         M4A  North York   
18     62         M5M  North York   
19     71         M6A  North York   
20     72         M6B  North York   
21     79         M6L  North York   
22     96         M9L  North York   
23     97         M9M  North York   

                                       Neighborhood   Latitude  Longitude  
0                                 Hillcrest Village  43.803762 -79.363452  
1                      Fairview, Henry Farm, Oriole  43.778517 -79.346556  
2                                   Bayview Village  43.786947 -79.385975  
4                           Newtonbrook, Willowdale  43.789053 -79.408493  
5                                  Willowdale South  43.770120 -79.408493  
6                                   York Mills West  43.752758 -79.400049  
7                                   Willowdale West  43.782736 -79.442259  
8                                         Parkwoods  43.753259 -79.329656  
9                                   Don Mills North  43.745906 -79.352188  
10                 Flemingdon Park, Don Mills South  43.725900 -79.340923  
11  Bathurst Manor, Downsview North, Wilson Heights  43.754328 -79.442259  
12                  Northwood Park, York University  43.767980 -79.487262  
13                      CFB Toronto, Downsview East  43.737473 -79.464763  
14                                   Downsview West  43.739015 -79.506944  
15                                Downsview Central  43.728496 -79.495697  
16                              Downsview Northwest  43.761631 -79.520999  
17                                 Victoria Village  43.725882 -79.315572  
18                Bedford Park, Lawrence Manor East  43.733283 -79.419750  
19                 Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
20                                        Glencairn  43.709577 -79.445073  
21               Downsview, North Park, Upwood Park  43.713756 -79.490074  
22                                    Humber Summit  43.756303 -79.565963  
23                                 Emery, Humberlea  43.724766 -79.532242

In [21]:
NorthYork.shape

(23, 6)

In [22]:
address = 'North York, Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of North York are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.20.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinates of North York are 43.7708175, -79.4132998.


Now we'll get ready to do some visualization with the folium package.

In [23]:
map_NorthYork = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(NorthYork['Latitude'], NorthYork['Longitude'], NorthYork['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_NorthYork)  
    
map_NorthYork

Alright, we're getting a good overview of North York, but let's dive further. We'll use the Foursquare API to get more details. The cell below contains Foursquare client data.

In [24]:
# @hidden cell

CLIENT_ID = 'X2P4NF4SR3MJ2XLNCOGV0OHNEWJV4GQIXJHOFWE0NXGN5MGM' # your Foursquare ID
CLIENT_SECRET = 'DHKELFYSLN2T3P4P4EDNPDSBD32FERWACHDS5F1VNX1R5U2K' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [25]:
NorthYork.loc[0,'Neighborhood']

'Hillcrest Village'

In [26]:
neighborhood_latitude = NorthYork.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = NorthYork.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = NorthYork.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Hillcrest Village are 43.8037622, -79.3634517.


In [27]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

The above url will get us all the details about the first neighborhood in North York.

In [28]:
results = requests.get(url).json()
#results

We have the results, but before we can do any true analysis we should clean it up so it's easier to understand.

In [29]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [30]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name                categories        lat        lng
0  Eagle's Nest Golf Club               Golf Course  43.805455 -79.364186
1         AY Jackson Pool                      Pool  43.804515 -79.366138
2            Villa Madina  Mediterranean Restaurant  43.801685 -79.363938
3       Duncan Creek Park                   Dog Run  43.805539 -79.360695

In [31]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


We've gotten the venues for a single neighborhood and everything looks good. Next step is to make the process work for every neighborhood in North York.

In [32]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [33]:
NorthYork_venues = getNearbyVenues(names=NorthYork['Neighborhood'],
                                   latitudes=NorthYork['Latitude'],
                                   longitudes=NorthYork['Longitude']
                                  )

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Bedford Park, Lawrence Manor East
Lawrence Heights, Lawrence Manor
Glencairn
Downsview, North Park, Upwood Park
Humber Summit
Emery, Humberlea


In [34]:
NorthYork_venues.shape

(246, 7)

In [35]:
NorthYork_venues.head()

Neighborhood  Neighborhood Latitude  \
0             Hillcrest Village              43.803762   
1             Hillcrest Village              43.803762   
2             Hillcrest Village              43.803762   
3             Hillcrest Village              43.803762   
4  Fairview, Henry Farm, Oriole              43.778517   

   Neighborhood Longitude                   Venue  Venue Latitude  \
0              -79.363452  Eagle's Nest Golf Club       43.805455   
1              -79.363452         AY Jackson Pool       43.804515   
2              -79.363452            Villa Madina       43.801685   
3              -79.363452       Duncan Creek Park       43.805539   
4              -79.346556          The LEGO Store       43.778207   

   Venue Longitude            Venue Category  
0       -79.364186               Golf Course  
1       -79.366138                      Pool  
2       -79.363938  Mediterranean Restaurant  
3       -79.360695                   Dog Run  
4       -79.343483          Toy / Game Store

Let's find out how many nearby venues are in each grouping of neighborhoods.

In [36]:
NorthYork_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Bathurst Manor, Downsview North, Wilson Heights                     19   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
CFB Toronto, Downsview East                                          3   
Don Mills North                                                      6   
Downsview Central                                                    4   
Downsview Northwest                                                  5   
Downsview West                                                       6   
Downsview, North Park, Upwood Park                                   4   
Emery, Humberlea                                                     2   
Fairview, Henry Farm, Oriole                                        69   
Flemingdon Park, Don Mills South                                    21   
Glencairn                                                            5   
Hillcrest Village                                                    4   
Humber Summit                                                        2   
Lawrence Heights, Lawrence Manor                                    10   
Northwood Park, York University                                      7   
Parkwoods                                                            2   
Victoria Village                                                     5   
Willowdale South                                                    35   
Willowdale West                                                      6   
York Mills West                                                      5   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Bathurst Manor, Downsview North, Wilson Heights                      19   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
CFB Toronto, Downsview East                                           3   
Don Mills North                                                       6   
Downsview Central                                                     4   
Downsview Northwest                                                   5   
Downsview West                                                        6   
Downsview, North Park, Upwood Park                                    4   
Emery, Humberlea                                                      2   
Fairview, Henry Farm, Oriole                                         69   
Flemingdon Park, Don Mills South                                     21   
Glencairn                                                             5   
Hillcrest Village                                                     4   
Humber Summit                                                         2   
Lawrence Heights, Lawrence Manor                                     10   
Northwood Park, York University                                       7   
Parkwoods                                                             2   
Victoria Village                                                      5   
Willowdale South                                                     35   
Willowdale West                                                       6   
York Mills West                                                       5   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Bathurst Manor, Downsview North, Wilson Heights     19              19   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
CFB Toronto, Downsview East                          3               3   
Don Mills North            

Now let's see how many unique venue categories there are in our data.

In [37]:
print('There are {} uniques categories.'.format(len(NorthYork_venues['Venue Category'].unique())))

There are 106 uniques categories.


In order to help us analyze the data proportionally, we'll use one-hot encoding.

In [38]:
NorthYork_onehot = pd.get_dummies(NorthYork_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
NorthYork_onehot['Neighborhood'] = NorthYork_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [NorthYork_onehot.columns[-1]] + list(NorthYork_onehot.columns[:-1])
NorthYork_onehot = NorthYork_onehot[fixed_columns]

NorthYork_onehot.head()

Neighborhood  Accessories Store  Airport  \
0             Hillcrest Village                  0        0   
1             Hillcrest Village                  0        0   
2             Hillcrest Village                  0        0   
3             Hillcrest Village                  0        0   
4  Fairview, Henry Farm, Oriole                  0        0   

   American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0                    0                    0                 0   
1                    0                    0                 0   
2                    0                    0                 0   
3                    0                    0                 0   
4                    0                    0                 0   

   Athletics & Sports  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0                   0       0     0    0               0                 0   
1                   0       0     0    0               0                 0   
2                   0       0     0    0               0                 0   
3                   0       0     0    0               0                 0   
4                   0       0     0    0               0                 0   

   Beer Store  Bike Shop  Boutique  Bridal Shop  Bubble Tea Shop  \
0           0          0         0            0                0   
1           0          0         0            0                0   
2           0          0         0            0                0   
3           0          0         0            0                0   
4           0          0         0            0                0   

   Burger Joint  Burrito Place  Bus Station  Butcher  Café  Candy Store  \
0             0              0            0        0     0            0   
1             0              0            0        0     0            0   
2             0              0            0        0     0            0   
3             0              0            0        0     0            0   
4             0              0            0        0     0            0   

   Caribbean Restaurant  Chinese Restaurant  Clothing Store  Coffee Shop  \
0                     0                   0               0            0   
1                     0                   0               0            0   
2                     0                   0               0            0   
3                     0                   0               0            0   
4                     0                   0               0            0   

   Comfort Food Restaurant  Concert Hall  Construction & Landscaping  \
0                        0             0                           0   
1                        0             0                           0   
2                        0             0                           0   
3                        0             0                           0   
4                        0             0                           0   

   Convenience Store  Cosmetics Shop  Deli / Bodega  Department Store  \
0                  0               0              0                 0   
1                  0               0              0                 0   
2                  0               0              0                 0   
3                  0               0              0                 0   
4                  0               0              0                 0   

   Dessert Shop  Dim Sum Restaurant  Diner  Discount Store  Dog Run  \
0             0                   0      0               0        0   
1             0                   0      0               0        0   
2             0                   0      0               0        0   
3             0                   0      0               0        1   
4             0                   0      0               0        0   

   Electronics Store  Empanada Restaurant  Falafel Restaurant  \
0                  0                    0                   0   
1                  0              

In [39]:
NorthYork_onehot.shape

(246, 107)

Now we'll group the rows by neighborhood and find the mean (or frequency) of each category's occurance.

In [40]:
NorthYork_grouped = NorthYork_onehot.groupby('Neighborhood').mean().reset_index()
NorthYork_grouped

Neighborhood  Accessories Store  \
0   Bathurst Manor, Downsview North, Wilson Heights           0.000000   
1                                   Bayview Village           0.000000   
2                 Bedford Park, Lawrence Manor East           0.000000   
3                       CFB Toronto, Downsview East           0.000000   
4                                   Don Mills North           0.000000   
5                                 Downsview Central           0.000000   
6                               Downsview Northwest           0.000000   
7                                    Downsview West           0.000000   
8                Downsview, North Park, Upwood Park           0.000000   
9                                  Emery, Humberlea           0.000000   
10                     Fairview, Henry Farm, Oriole           0.014493   
11                 Flemingdon Park, Don Mills South           0.000000   
12                                        Glencairn           0.000000   
13                                Hillcrest Village           0.000000   
14                                    Humber Summit           0.000000   
15                 Lawrence Heights, Lawrence Manor           0.100000   
16                  Northwood Park, York University           0.000000   
17                                        Parkwoods           0.000000   
18                                 Victoria Village           0.000000   
19                                 Willowdale South           0.000000   
20                                  Willowdale West           0.000000   
21                                  York Mills West           0.000000   

     Airport  American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0   0.000000             0.000000             0.000000          0.000000   
1   0.000000             0.000000             0.000000          0.000000   
2   0.000000             0.045455             0.000000          0.000000   
3   0.333333             0.000000             0.000000          0.000000   
4   0.000000             0.000000             0.000000          0.000000   
5   0.000000             0.000000             0.000000          0.000000   
6   0.000000             0.000000             0.000000          0.000000   
7   0.000000             0.000000             0.000000          0.000000   
8   0.000000             0.000000             0.000000          0.000000   
9   0.000000             0.000000             0.000000          0.000000   
10  0.000000             0.014493             0.000000          0.028986   
11  0.000000             0.000000             0.000000          0.095238   
12  0.000000             0.000000             0.000000          0.000000   
13  0.000000             0.000000             0.000000          0.000000   
14  0.000000             0.000000             0.000000          0.000000   
15  0.000000             0.000000             0.000000          0.000000   
16  0.000000             0.000000             0.000000          0.000000   
17  0.000000             0.000000             0.000000          0.000000   
18  0.000000             0.000000             0.000000          0.000000   
19  0.000000             0.000000             0.028571          0.000000   
20  0.000000             0.000000             0.000000          0.000000   
21  0.000000             0.000000             0.000000          0.000000   

    Athletics & Sports    Bakery      Bank       Bar  Baseball Field  \
0                  0.0  0.000000  0.052632  0.000000        0.000000   
1                  0.0  0.000000  0.250000  0.000000        0.000000   
2                  0.0  0.000000  0.000000  0.000000        0.000000   
3                  0.0  0.000000  0.000000  0.000000        0.000000   
4                  0.0  0.000000  0.000000  0.000000        0.166667   
5                  0.0  0.000000  0.000000  0.000000        0.250000   
6                  0.2  0.000000  0.000000  0.000000        0.000000   
7          

Using this dataframe, we can find the top 5 most common venue types for each grouping of neighborhoods.

In [41]:
num_top_venues = 5

for hood in NorthYork_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = NorthYork_grouped[NorthYork_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor, Downsview North, Wilson Heights----
                venue  freq
0         Coffee Shop  0.11
1    Sushi Restaurant  0.05
2  Chinese Restaurant  0.05
3      Sandwich Place  0.05
4       Shopping Mall  0.05


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Bank  0.25
2  Japanese Restaurant  0.25
3                 Café  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                  venue  freq
0    Italian Restaurant  0.09
1           Coffee Shop  0.09
2     Indian Restaurant  0.05
3      Sushi Restaurant  0.05
4  Fast Food Restaurant  0.05


----CFB Toronto, Downsview East----
                        venue  freq
0                     Airport  0.33
1                        Park  0.33
2  Construction & Landscaping  0.33
3                Home Service  0.00
4                       Hotel  0.00


----Don Mills North----
                  venue  freq
0  Gym / Fitness Center  0.17
1  Caribbean

Let's put this information into a new dataframe so that it will easier to visualize later.

In [42]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = NorthYork_grouped['Neighborhood']

for ind in np.arange(NorthYork_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(NorthYork_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood  \
0   Bathurst Manor, Downsview North, Wilson Heights   
1                                   Bayview Village   
2                 Bedford Park, Lawrence Manor East   
3                       CFB Toronto, Downsview East   
4                                   Don Mills North   
5                                 Downsview Central   
6                               Downsview Northwest   
7                                    Downsview West   
8                Downsview, North Park, Upwood Park   
9                                  Emery, Humberlea   
10                     Fairview, Henry Farm, Oriole   
11                 Flemingdon Park, Don Mills South   
12                                        Glencairn   
13                                Hillcrest Village   
14                                    Humber Summit   
15                 Lawrence Heights, Lawrence Manor   
16                  Northwood Park, York University   
17                                        Parkwoods   
18                                 Victoria Village   
19                                 Willowdale South   
20                                  Willowdale West   
21                                  York Mills West   

            1st Most Common Venue 2nd Most Common Venue  \
0                     Coffee Shop   Fried Chicken Joint   
1              Chinese Restaurant                  Café   
2                     Coffee Shop    Italian Restaurant   
3                            Park               Airport   
4             Japanese Restaurant  Gym / Fitness Center   
5                    Home Service            Food Truck   
6                   Grocery Store  Gym / Fitness Center   
7                   Grocery Store                 Hotel   
8      Construction & Landscaping                Bakery   
9   Paper / Office Supplies Store        Baseball Field   
10                 Clothing Store           Coffee Shop   
11                     Beer Store           Coffee Shop   
12                         Bakery                   Pub   
13                    Golf Course               Dog Run   
14                    Pizza Place   Empanada Restaurant   
15         Furniture / Home Store     Accessories Store   
16                    Coffee Shop    Miscellaneous Shop   
17                           Park     Food & Drink Shop   
18                    Coffee Shop           Pizza Place   
19                    Coffee Shop      Ramen Restaurant   
20                  Grocery Store        Discount Store   
21                           Park     Convenience Store   

         3rd Most Common Venue     4th Most Common Venue  \
0                Shopping Mall        Chinese Restaurant   
1                         Bank       Japanese Restaurant   
2                Grocery Store         Indian Restaurant   
3   Construction & Landscaping         Electronics Store   
4         Caribbean Restaurant                      Café   
5               Baseball Field         Korean Restaurant   
6           Athletics & Sports            Discount Store   
7                         Bank                      Park   
8               Massage Studio                      Park   
9                Women's Store         Electronics Store   
10        Fast Food Restaurant             Women's Store   
11                         Gym          Asian Restaurant   
12         Filipino Restaurant        Italian Restaurant   
13                        Pool  Mediterranean Restaurant   
14               Women's Store                   Dog Run   
15          Athletics & Sports        Miscellaneous Shop   
16        Caribbean Restaurant             Metro Station   
17                     Dog Run              Concert Hall   
18                Hockey Arena     Portuguese Restaurant   
19                 Pizza Place       Japanese Restaurant   
20                    Pharmacy               Pizza Place   
21                        Bank                       Bar   

         5th Most Common Venue    

## Clustering

Let's cluster North York into five groups.

In [44]:
# set number of clusters
kclusters = 5

NorthYork_grouped_clustering = NorthYork_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(NorthYork_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 1, 0, 2, 0, 0, 1, 2], dtype=int32)

In [45]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

NorthYork_merged = NorthYork

Now we can merge the clusters with the North York dataframe we created earlier.

In [46]:
# add clustering labels

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
NorthYork_merged = NorthYork_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

NorthYork_merged.head() # check the last columns!

index Postal Code     Borough                  Neighborhood   Latitude  \
0     17         M2H  North York             Hillcrest Village  43.803762   
1     18         M2J  North York  Fairview, Henry Farm, Oriole  43.778517   
2     19         M2K  North York               Bayview Village  43.786947   
4     21         M2M  North York       Newtonbrook, Willowdale  43.789053   
5     22         M2N  North York              Willowdale South  43.770120   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.363452             3.0           Golf Course               Dog Run   
1 -79.346556             0.0        Clothing Store           Coffee Shop   
2 -79.385975             0.0    Chinese Restaurant                  Café   
4 -79.408493             NaN                   NaN                   NaN   
5 -79.408493             0.0           Coffee Shop      Ramen Restaurant   

  3rd Most Common Venue     4th Most Common Venue       5th Most Common Venue  \
0                  Pool  Mediterranean Restaurant  Construction & Landscaping   
1  Fast Food Restaurant             Women's Store                      Bakery   
2                  Bank       Japanese Restaurant               Women's Store   
4                   NaN                       NaN                         NaN   
5           Pizza Place       Japanese Restaurant                        Café   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Convenience Store        Cosmetics Shop         Deli / Bodega   
1   Sporting Goods Shop   Japanese Restaurant     Electronics Store   
2     Electronics Store     Convenience Store        Cosmetics Shop   
4                   NaN                   NaN                   NaN   
5        Sandwich Place      Sushi Restaurant                 Hotel   

  9th Most Common Venue 10th Most Common Venue  
0      Department Store           Dessert Shop  
1      Asian Restaurant         Cosmetics Shop  
2         Deli / Bodega       Department Store  
4                   NaN                    NaN  
5       Bubble Tea Shop                  Plaza

Let's map this information.

In [48]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)-1))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi in zip(NorthYork_merged['Latitude'], NorthYork_merged['Longitude'], NorthYork_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Clusters ' + str(kclusters), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Finally, we can examine a cluster to see the details regarding it's venues.

In [50]:
NorthYork_merged.loc[NorthYork_merged['Cluster Labels'] == 0, NorthYork_merged.columns[[1] + list(range(5, NorthYork_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels   1st Most Common Venue  \
1          M2J -79.346556             0.0          Clothing Store   
2          M2K -79.385975             0.0      Chinese Restaurant   
5          M2N -79.408493             0.0             Coffee Shop   
7          M2R -79.442259             0.0           Grocery Store   
9          M3B -79.352188             0.0     Japanese Restaurant   
10         M3C -79.340923             0.0              Beer Store   
11         M3H -79.442259             0.0             Coffee Shop   
12         M3J -79.487262             0.0             Coffee Shop   
14         M3L -79.506944             0.0           Grocery Store   
16         M3N -79.520999             0.0           Grocery Store   
17         M4A -79.315572             0.0             Coffee Shop   
18         M5M -79.419750             0.0             Coffee Shop   
19         M6A -79.464763             0.0  Furniture / Home Store   
20         M6B -79.445073             0.0                  Bakery   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1            Coffee Shop  Fast Food Restaurant          Women's Store   
2                   Café                  Bank    Japanese Restaurant   
5       Ramen Restaurant           Pizza Place    Japanese Restaurant   
7         Discount Store              Pharmacy            Pizza Place   
9   Gym / Fitness Center  Caribbean Restaurant                   Café   
10           Coffee Shop                   Gym       Asian Restaurant   
11   Fried Chicken Joint         Shopping Mall     Chinese Restaurant   
12    Miscellaneous Shop  Caribbean Restaurant          Metro Station   
14                 Hotel                  Bank                   Park   
16  Gym / Fitness Center    Athletics & Sports         Discount Store   
17           Pizza Place          Hockey Arena  Portuguese Restaurant   
18    Italian Restaurant         Grocery Store      Indian Restaurant   
19     Accessories Store    Athletics & Sports     Miscellaneous Shop   
20                   Pub   Filipino Restaurant     Italian Restaurant   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1                      Bakery   Sporting Goods Shop   Japanese Restaurant   
2               Women's Store     Electronics Store     Convenience Store   
5                        Café        Sandwich Place      Sushi Restaurant   
7                     Butcher           Coffee Shop            Food Truck   
9              Baseball Field      Basketball Court   Empanada Restaurant   
10         Chinese Restaurant        Clothing Store    Dim Sum Restaurant   
11  Middle Eastern Restaurant    Frozen Yogurt Shop              Pharmacy   
12             Massage Studio                   Bar    Falafel Restaurant   
14              Shopping Mall               Dog Run          Concert Hall   
16               Liquor Store     Electronics Store     Convenience Store   
17               Intersection        Discount Store          Concert Hall   
18               Liquor Store        Cosmetics Shop             Juice Bar   
19             Clothing Store              Boutique           Coffee Shop   
20        Japanese Restaurant         Women's Store               Dog Run   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1            Electronics Store      Asian Restaurant         Cosmetics Shop  
2               Cosmetics Shop         Deli / Bodega       Department Store  
5                        Hotel       Bubble Tea Shop                  Plaza  
7                 Dessert Shop          Concert Hall              Gift Shop  
9               Cosmetics Shop         Deli / Bodega       Department Store  
10              Discount Store   Japanese Restaurant     Italian Restaurant  
11                 Pizza Place         Deli / Bodega            Bridal Shop  
12               Women's Store        Discount Store         Cosmetics Shop  
14  Construct

Using this information, we could make inferences about people's interests in each area of the city. We could also compare clusters between boroughs of Toronto or even another city to determine similarities.